File: ruby/palindrome-products/palindrome_products.rb
class Palindromes

  attr_reader :min, :max, :palindromes

  def initialize(options = {})
    @min = options[:min_factor] || 1
    @max = options[:max_factor] || 10
    @palindromes = Hash.new { |hash,key| hash[key] = [] }
  end

  def generate
    candidates.select { |a,b| palindrome?(a,b) }.each do |a,b|
      palindromes[a * b] << [a,b]
    end
  end

  def smallest
    key = palindromes.keys.min
    Palindrome.new(key, palindromes[key])
  end

  def largest
    key = palindromes.keys.max
    Palindrome.new(key, palindromes[key])
  end

  private

  def candidates
    (min..max).to_a.combination(2).to_a + (min..max).map { |num| [num,num] }
  end

  def palindrome?(a, b)
    str = (a * b).to_s
    str == str.reverse
  end

  Palindrome = Struct.new(:value, :factors)

end

File: elixir/palindrome-products/palindrome_products.exs
defmodule Palindromes do

  @doc """
  Generates all palindrome products from an optionally given min factor (or 1) to a given max factor.
  """
  @spec generate(non_neg_integer, non_neg_integer) :: map
  def generate(max_factor, min_factor \\ 1) do
    palindromes(min_factor, max_factor)
    |> Enum.group_by(fn([a,b]) -> a*b end)
  end

  defp palindromes(min, max) do
    for a <- min..max, b <- a..max, palindrome?(a*b), do: [a,b]
  end

  defp palindrome?(num) do
    n = Integer.to_string(num)
    n == String.reverse(n)
  end

end
